[Static Analyzer CI] Update the summary page for unexpected results

briannafan opened the pull request:
#### b1846eb8836ac04b508679d16a0ed048cf3c63d0
<pre>
[Static Analyzer CI] Update the summary page for unexpected results
<a href="https://bugs.webkit.org/show_bug.cgi?id=279465">https://bugs.webkit.org/show_bug.cgi?id=279465</a>
<a href="https://rdar.apple.com/135748287">rdar://135748287</a>

Reviewed by Ryosuke Niwa.

Update the UI of the unexpected results page.
Now results are sorted by passes and failures, then by project.
File lists and counts are displayed in a table.

* Tools/Scripts/generate-static-analysis-archive:
(parse_command_line):
(generate_unexpected_table):
(generate_unexpected_results_page):
(generate_results_page):
(create_results_file):
(generate_unexpected_results_links): Deleted.

Canonical link: <a href="https://commits.webkit.org/283445@main">https://commits.webkit.org/283445@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4936be304c1cf95fdc1916cf42966c089c15229c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66341 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/45716 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/18962 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70373 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/16952 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/68459 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/53515 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/17234 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/53206 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/59/builds/16952 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69408 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/53515 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/57425 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/33857 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/53515 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/14810 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/15827 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/53515 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/15153 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72077 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10298 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/17234 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/60532 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10330 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/57492 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/60842 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/88/builds/8497 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/2112 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10046 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/41523 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/42600 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/43783 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/42343 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Tools/Scripts/generate-static-analysis-archive:
```diff
@@ -58,30 +101,49 @@ def parse_command_line(args):
+            project_count = get_project_issue_count_as_string(static_analyzer_index_path)
+            body += f'<h3>{project_name}</h3>'
+            if int(project_count) != 0 and category == 'UnexpectedFailures':
+                body += f"<b>{LINK_TEMPLATE.format(link=os.path.relpath(static_analyzer_index_path, output_root), text=f'View Issues ({project_count})')}</b></br></br>"
```
As I commented in slack, we may want to omit this number since it's rather confusing to have two different kinds of numbers in the results page.

briannafan replied:
Agreed

rniwa commented on Tools/Scripts/generate-static-analysis-archive:
```diff
@@ -33,6 +33,7 @@ INDEX_TEMPLATE = """
 <html>
 <head>
     <title>{title}</title>
+    <style>body {{ font-family: Helvetica, sans-serif; font-size:10pt }}</style>
```
Nit: need a space after `font-size:`.

rniwa commented on Tools/Scripts/generate-static-analysis-archive:
```diff
@@ -42,10 +43,52 @@ INDEX_TEMPLATE = """
+    <style>
+        body {{ font-family: Helvetica, sans-serif; font-size:10pt }}
+        table {{ font-size:10pt }}
+        table {{ border-spacing: 0px; border: 1px solid black}}
```
NIt: Need a space between `lack` and `}}`

rniwa commented on Tools/Scripts/generate-static-analysis-archive:
```diff
@@ -42,10 +43,52 @@ INDEX_TEMPLATE = """
+        table {{ font-size:10pt }}
+        table {{ border-spacing: 0px; border: 1px solid black}}
+        th, table thead {{
+        background-color:#eee; color:#666666;
```
We usually indent CSS rules like these.

rniwa commented on Tools/Scripts/generate-static-analysis-archive:
```diff
@@ -58,30 +101,49 @@ def parse_command_line(args):
-    return results_links
+def generate_unexpected_table(output_root, static_analysis_dir, project_name, category):
+    # Get lists of unexpected results per project
+    category_table = ''
```
Rather than adding a comment, we should rename this variable to unexpected_results_per_project.

rniwa commented on Tools/Scripts/generate-static-analysis-archive:
```diff
@@ -58,30 +101,49 @@ def parse_command_line(args):
+        return -1
+    unexpected_results_files = result_files.splitlines()
+
+    # Add links to unexpected lists and counts in a table
```
Rather than adding a comment like this, we may want to extract the below code as a helper method like this:
```
table_rows = create_table_for_unexpected_results_per_project(unexpected_results_per_project)
```

briannafan committed:
[Static Analyzer CI] Update the summary page for unexpected results
https://bugs.webkit.org/show_bug.cgi?id=279465
rdar://135748287

Reviewed by Ryosuke Niwa.

Update the UI of the unexpected results page.
Now results are sorted by passes and failures, then by project.
File lists and counts are displayed in a table.

* Tools/Scripts/generate-static-analysis-archive:
(parse_command_line):
(generate_unexpected_table):
(generate_unexpected_results_page):
(generate_results_page):
(create_results_file):
(generate_unexpected_results_links): Deleted.

Canonical link: https://commits.webkit.org/283445@main
